gstreamer/src/query.rs

```diff
@@ -15,188 +15,188 @@ use std::ptr;
+        )
     }
 
-    pub fn is_upstream(&self) -> bool {
-        unsafe {
-            ((*self.as_ptr()).type_ as u32) & (ffi::GST_QUERY_TYPE_UPSTREAM.bits()) != 0
-        }
+    pub fn new_accept_caps(caps: &::Caps) -> AcceptCaps {
+        assert_initialized_main_thread!();
+        AcceptCaps(
+            unsafe { from_glib_full(ffi::gst_query_new_accept_caps(caps.as_mut_ptr())) }
+        )
     }
 
-    pub fn is_serialized(&self) -> bool {
-        unsafe {
-            ((*self.as_ptr()).type_ as u32) & (ffi::GST_QUERY_TYPE_SERIALIZED.bits()) != 0
-        }
+    pub fn new_caps<'a, P: Into<Option<&'a ::Caps>>>(filter: P) -> Caps {
+        assert_initialized_main_thread!();
+        let filter = filter.into();
+        Caps(
+            unsafe { from_glib_full(ffi::gst_query_new_caps(filter.to_glib_none().0)) }
+        )
     }
 
-    pub fn view(&self) -> QueryView<&Self> {
-        let type_ = unsafe { (*self.as_ptr()).type_ };
+    pub fn new_drain() -> Drain {
+        assert_initialized_main_thread!();
+        Drain(
+            unsafe { from_glib_full(ffi::gst_query_new_drain()) }
+        )
+    }
 
-        match type_ {
-            ffi::GST_QUERY_POSITION => QueryView::Position(Position(self)),
-            ffi::GST_QUERY_DURATION => QueryView::Duration(Duration(self)),
-            ffi::GST_QUERY_LATENCY => QueryView::Latency(Latency(self)),
-            ffi::GST_QUERY_JITTER => QueryView::Jitter(Jitter(self)),
-            ffi::GST_QUERY_RATE => QueryView::Rate(Rate(self)),
-            ffi::GST_QUERY_SEEKING => QueryView::Seeking(Seeking(self)),
-            ffi::GST_QUERY_SEGMENT => QueryView::Segment(Segment(self)),
-            ffi::GST_QUERY_CONVERT => QueryView::Convert(Convert(self)),
-            ffi::GST_QUERY_FORMATS => QueryView::Formats(Formats(self)),
-            ffi::GST_QUERY_BUFFERING => QueryView::Buffering(Buffering(self)),
-            ffi::GST_QUERY_CUSTOM => QueryView::Custom(Custom(self)),
-            ffi::GST_QUERY_URI => QueryView::Uri(Uri(self)),
-            ffi::GST_QUERY_ALLOCATION => QueryView::Allocation(Allocation(self)),
-            ffi::GST_QUERY_SCHEDULING => QueryView::Scheduling(Scheduling(self)),
-            ffi::GST_QUERY_ACCEPT_CAPS => QueryView::AcceptCaps(AcceptCaps(self)),
-            ffi::GST_QUERY_CAPS => QueryView::Caps(Caps(self)),
-            ffi::GST_QUERY_DRAIN => QueryView::Drain(Drain(self)),
-            ffi::GST_QUERY_CONTEXT => QueryView::Context(Context(self)),
-            _ => QueryView::Other(Other(self)),
-        }
-    }
-
-    pub fn view_mut(&mut self) -> QueryView<&mut Self> {
-        unsafe { mem::transmute(self.view()) }
+    pub fn new_context(context_type: &str) -> Context {
+        assert_initialized_main_thread!();
+        Context(
+            unsafe { from_glib_full(ffi::gst_query_new_context(context_type.to_glib_none().0)) }
+        )
     }
 }
 
-impl glib::types::StaticType for QueryRef {
+impl glib::types::StaticType for Query {
```

Comment:
With this change it would be possible to store queries in GValues, but that can probably be added again somehow